storage/innobase/btr/btr0cur.cc

```diff
@@ -1331,35 +1348,48 @@ dberr_t btr_cur_t::search_leaf(const dtuple_t *tuple, page_cur_mode_t mode,
       We may have to reacquire the page latch in a different mode. */
       switch (rw_latch) {
       case RW_S_LATCH:
-        if ((latch_mode & ~12) != RW_S_LATCH)
+        if (!(latch_mode & BTR_SEARCH_LEAF))
         {
+          rw_latch= RW_X_LATCH;
           ut_ad(rw_lock_type_t(latch_mode & ~12) == RW_X_LATCH);
-          goto relatch_x;
-        }
-        if (latch_mode != BTR_MODIFY_PREV)
-        {
-          if (!latch_by_caller)
-            /* Release the tree s-latch */
-            mtr->rollback_to_savepoint(savepoint, savepoint + 1);
-          goto reached_latched_leaf;
+          mtr->lock_register(block_savepoint, MTR_MEMO_PAGE_X_FIX);
+          if (!block->page.lock.s_x_upgrade_try())
+          {
+            block->page.lock.s_unlock();
+            block->page.lock.x_lock();
+            /* Dropping the index tree (and freeing the root page)
+            should be impossible while we hold index()->lock. */
+            ut_ad(!block->page.is_freed());
+            page_level= btr_page_get_level(block->page.frame);
+            if (UNIV_UNLIKELY(page_level != 0))
+            {
+              /* btr_root_raise_and_insert() was executed meanwhile */
+              ut_ad(mtr->memo_contains_flagged(&index()->lock,
+                                               MTR_MEMO_S_LOCK));
+              goto reached_root;
+            }
+          }
         }
-        /* fall through */
+        if (latch_mode == BTR_MODIFY_PREV)
+          goto reached_leaf;
+        if (!latch_by_caller)
+          /* Release the tree s-latch */
+          mtr->rollback_to_savepoint(savepoint, savepoint + 1);
+        goto reached_latched_leaf;
```

Comment:
We would got to "reached_latched_leaf" after upgrading to X. AFAICS, the flow earlier for the same case would go to `reached_root_and_leaf` after releasing S and re-acquiring X lock. There are differences in the paths e.g. we have calls like btr_search_drop_page_hash_index() for to get rid of AHI for the page being updated etc.

```
        if (latch_mode == BTR_MODIFY_TREE)
          goto reached_index_root_and_leaf;
        goto reached_root_and_leaf;
```
I haven't checked if this could introduce any real issue but to eliminate such possibilities, we could  set a label at `case RW_X_LATCH:` and jump directly there after upgrading to X lock and also for `latch_mode == BTR_MODIFY_PREV`. Since our primary goal is to replace the release + re-acquire to lock upgrade it is better to keep the rest of the flow same. What do you think ?

Reply:
The question basically boils down to: Why are we not doing `goto reached_root_and_leaf` in the `case RW_S_LATCH:` branch of `switch (rw_latch)`, like we do in all other branches? There used to be a `goto reached_leaf` after the `switch (rw_latch)` block. Now I realize that because there was no `break` in any of the `case:` branches, that `goto reached_leaf` was actually unreachable. I agree that we should just replace the `goto reached_leaf` after the `switch (rw_latch)` with `goto reached_root_and_leaf`, because it is the same state in each case: we are holding an exclusive latch on the single root-and-leaf page of the index.

Reply:
If `latch_mode==BTR_SEARCH_LEAF`, we do not want to get here:
```c++
    reached_index_root_and_leaf:
      ut_ad(rw_latch == RW_X_LATCH);
```
We can avoid that with the following:
```c++
if (rw_latch == RW_S_LATCH)
  goto reached_latched_leaf;
// …
goto reached_root_and_leaf;
```

Reply:
Addressed in 418147edc388f82e452c9bc93d686de9027ad219.